tsne/bh_sne_src/tsne.cpp

```diff
@@ -706,16 +714,60 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
             "%f)!\nLearning embedding...\n",
             (float)(end - start) / CLOCKS_PER_SEC,
             (double)row_P[N] / ((double)N * (double)N));
-  start = clock();
 
-  for (int iter = 0; iter < max_iter; iter++) {
+  TSNEState* state = new TSNEState;
+  state->P = P;
+  state->row_P = row_P;
+  state->col_P = col_P;
+  state->val_P = val_P;
+  state->dY = dY;
+  state->uY = uY;
+  state->gains = gains;
+  TSNE* tsne = new TSNE;
+  tsne->N = N;
+  tsne->Y = Y;
+  tsne->no_dims = NDIMS;
+  tsne->theta = theta;
+  tsne->max_iter = max_iter;
+  tsne->stop_lying_iter = stop_lying_iter;
+  tsne->mom_switch_iter = mom_switch_iter;
+  tsne->state = state;
+  tsne->iter = 0;
+  tsne->total_time = 0;
+  tsne->residual_time = 0;
+
+  return tsne;
+}
+
+// Optimize t-SNE
+template <int NDIMS>
+bool run_n(int n, struct TSNE* tsne) {
+  // Set learning parameters
+  double momentum = .5, final_momentum = .8;
+  double eta = 200.0;
+
+  // Extract state
+  int N = tsne->N;
+  double* Y = tsne->Y;
+  double theta = tsne->theta;
+  bool exact = (theta == .0) ? true : false;
+  TSNEState* state = tsne->state;
+  vector<double>& dY = state->dY;
+  vector<double>& uY = state->uY;
```

Comment:
If you're going to do this, you might as well also do it for `P`, `row_P`, `col_P`, and `val_P` to avoid the diffs later on.

Reply:
Yeah, you're right.